File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.py
def solution(message, spoiler_ranges):
    words = message.split(" ")    
    start_idx = 0
    word_info = [["", 0, 0] for _ in range(len(words))]
    spoiled_range = [False for _ in range(len(message))]
    
    for i in range(len(words)):
        word = words[i]
        end_idx = start_idx + len(word) - 1
        word_info[i][0] = word
        word_info[i][1] = start_idx
        word_info[i][2] = end_idx
        start_idx = end_idx + 2
        
    for spoil_start, spoil_end in spoiler_ranges:
        spoiled_range[spoil_start:spoil_end + 1] = [True] * (spoil_end - spoil_start + 1)                
            
    spoiler_words = set()    
    no_spoiler_words = set()
    # 한 번만 등장산 스포 방지 단어가 중요 단어
    important_words = set()
    
    for word, start, end in word_info:
        is_spoiler = False
        for i in range(start, end + 1):
            if spoiled_range[i]:
                is_spoiler = True
                break
        if is_spoiler and not word in spoiler_words:
            spoiler_words.add(word)        
        elif not is_spoiler:
            no_spoiler_words.add(word)        
    
    important_words = spoiler_words - no_spoiler_words
    # print(spoiler_words)
    # print(no_spoiler_words)
    # print(important_words)
    
    answer = len(important_words)
    return answer

File: 프로그래머스/1/468370. 중요한 단어를 스포 방지/중요한 단어를 스포 방지.java
import java.util.*;

class Solution {
    public int solution(String message, int[][] spoiler_ranges) {
        String[] words = message.split(" ");
        boolean[] spoilerRange = new boolean[message.length() + 1];
        int[][] wordInfo = new int[words.length][2];    // 단어의 시작과 끝
        
        for(int[] spoiler: spoiler_ranges) {
            int start = spoiler[0];
            int end = spoiler[1];
            
            for(int i = start; i <= end; i++) {
                spoilerRange[i] = true;
            }
        }
        
        // 단어별로 시작위치, 끝나는 위치 설정
        int wordStart = 0;        
        for(int i = 0; i < words.length; i++){
            int wordEnd = wordStart + words[i].length() - 1;
            wordInfo[i][0] = wordStart;
            wordInfo[i][1] = wordEnd;
            wordStart = wordEnd + 2;
        }
                
        // 각각 스포일러 방지 단어, 중요하지 않은 단어, 이전에 등장했던 단어(스포방지 구간 외)
        HashSet<String> spoilerAlertWords = new HashSet<>();
        HashSet<String> notImportantWords = new HashSet<>();        
                
        // System.out.println(Arrays.toString(words));
        // System.out.println(Arrays.toString(spoilerRange));
        // System.out.println(Arrays.deepToString(wordInfo));
        
        for(int i = 0; i < words.length; i++) {
            boolean isSpoilerWord = false;
            for(int j = wordInfo[i][0]; j <= wordInfo[i][1]; j++){                
                if(spoilerRange[j]) {                    
                    spoilerAlertWords.add(words[i]);                    
                    isSpoilerWord = true;
                    break;
                }
            }
            
            if(!isSpoilerWord) {
                notImportantWords.add(words[i]);
            }
        }
        
        // System.out.println(spoilerAlertWords);
        // System.out.println(notImportantWords);
        
        spoilerAlertWords.removeAll(notImportantWords);        
        return spoilerAlertWords.size();
    }
}